### PR TITLE
feat(link_to): Add support for linking to existing files from the cache

### PR DESCRIPTION
**Pull request by shaug**
Hi, @zkat!

Thanks for this great project. It *almost* solves everything I need in a file-based cache. 😁 

For various reasons outside my control, I can't use the existing `cacache::Writer` functionality for some of the content in my app, but I do want these files to be indexed by a named 'key' and have proper integrity checking when accessed. And the `cacache::write` functionality will be useful for other types of content that I'm working with, so extending `cacache` to support symlinking seemed like a good way to handle this.

This PR addresses this by adding a new `link` feature to the project, which (when enabled) adds new APIs for symlinking existing files into the cache *without* copying over their data. It adds the following top-level functions:

- `cacache::link`
- `cacache::link_hash`
- `cacache::link_sync`
- `cacache::link_hash_sync`

It also adds the `cacache::Linker` and `cacache::SyncLinker` structs, which implements `AsyncRead` and `Read`, respectively. These classes are an interesting combination of both read and write functionality, as they need to read in the contents of the linked file to generate the `Integrity`, but also provide a `commit()` function to finalize writing index files (if a key is specified) and (symlinked) cache entries.

To enable this without a ton of duplication, the `WriteOpts` struct was extended with corresponding 'linking' functions. The symlinking code is portable across unix and windows.

The code is fully tested and benchmarked, following the patterns already established in the project, and I updated the docs everywhere it seemed relevant.

If you think linking external files is outside the scope of what you want for this project, that's totally fine. I just wanted to offer it as a contribution back in case you think it's worthwhile including in the 'main' cacache project.

Thanks for considering it!

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "cacache"
-version = "11.1.1-alpha.0"
+version = "11.2.0-alpha.0"
```
Comment (zkat) [inline Cargo.toml:3]: you can leave this alone. It's taken care of by release scripts.

**Review by zkat — COMMENTED** [review]
Hi! Thanks for this!

This definitely seems like a pretty neat feature. I won't pry too much but I'm curious about what scenario entailed you not being able to use cacache as the "authorative" store, though.

My main concern with this PR as-is is that I think there's two use-cases for linking involved with cacache, and this PR doesn't make clear which one that is (and I thought it was the OTHER one, at first, until I read more carefully):

1. Linking existing data _into_ cacache, to prevent copying when writing into it, but making the data available as regular cache data on read.
2. (Sym/hard)-linking _from_ cacache to a destination, instead of doing a full file read.

In fact, I had been thinking recently that I wanted use case 2 and was thinking of adding it myself.

So I guess my concern is: how can we (re)name the feature in this PR such that it's clear that this is the first use-case, and while leaving open the door for use case 2 in the future, without confusing people as to which is which?

**Comment by zkat** [comment]
I was thinking something along the lines of `link_into` and `link_from`? wdyt?

**Comment by shaug** [comment]
> I was thinking something along the lines of link_into and link_from? wdyt?

I can make `link_into` work, but I'm not sure `link_from` provides enough clarity that it is linking 'out' from the cache. The 'from' could easily be interpreted as 'from' the file being brought 'into' the cache.

I might consider `link_away`; I think that provides stronger directional clarity. Alternatively, I think `link_in` vs `link_out` might be less ambiguous (that said, I don't think `link_into` is ambiguous, but `link_in` is a better counter-name to `link_out`, if you prefer this alternative).

In the end, the `link_from`/`link_away`/`link_out` feature is yours to name. I'll be happy with whatever pairing you prefer, and I'll start the appropriate changes to my PR once you make your decree.

**Comment by zkat** [comment]
Let's do ~~LinkedIn~~`link-in` and `link-out`, then. I think those are both reasonable :)

**Comment by shaug** [comment]
onnit!

**Comment by shaug** [comment]
Phew! So katz says, so it shall be.

No more naming suggestions from me, I promise. This is the right one.

**Comment by shaug** [comment]
OK! I think it has been fully `link_to`ified.

File: src/content/linkto.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+use std::fs::File;
+use std::path::{Path, PathBuf};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use crate::async_lib::AsyncRead;
+use crate::content::path;
+use crate::errors::{IoErrorExt, Result};
+
+#[cfg(not(any(unix, windows)))]
```
Comment (zkat) [inline src/content/linkto.rs:12]: I thought the correct way to do this was `target_family = "unix"/"windows"`. I'm not sure this conditional does anything. Does it?? I guess it must've compiled for you, so...

Comment (shaug) [reply src/content/linkto.rs:12]: https://doc.rust-lang.org/reference/conditional-compilation.html#unix-and-windows

> `unix` is set if `target_family = "unix"` is set and `windows` is set if `target_family = "windows"` is set.

Happy to use the more explicit `target_family` if you prefer

Comment (zkat) [reply src/content/linkto.rs:12]: nope, I'm happy with this :) Just making sure I was getting it right. I've been having a lot of Adventures™️ with cfg recently :P

**Comment by shaug** [comment]
Awesome! Thanks, @zkat!

**Comment by zkat** [comment]
This has been released as part of `11.2.0`, along with #42, which might also be of interest!

**Pull request merged**